### ISSUE TITLE
Wrong memory order with pin_memory() over transposed tensors

### ISSUE THREAD

**Issue by ozancaglayan** [pytorch/pytorch#3266]
Hello,

I was experimenting with a custom `collate_fn` which receives a batch of different sized integer sequences and pads them to form 2D tensors. I noticed a weird issue for which I created the following MWE:

```python
import torch

t = torch.FloatTensor(range(10)).view((2,5))
print(t.t())
print(t.t().pin_memory())

# Output
 0  5
 1  6
 2  7
 3  8
 4  9
[torch.FloatTensor of size 5x2]

 0  1
 2  3
 4  5
 6  7
 8  9
[torch.FloatTensor of size 5x2]
```

It's probably because of the different memory ordering on Python's side and CUDA's allocator.
Tested on K40 and GTX1080 with driver 375.39, CUDA8.

**Comment by ozancaglayan** [comment]
Seems to be related to column/row ordering stuff:

```
In [19]: x = np.array(range(10)).reshape((2,5))

In [20]: torch.from_numpy(x.T)
Out[20]:

 0  5
 1  6
 2  7
 3  8
 4  9
[torch.LongTensor of size 5x2]

In [21]: torch.from_numpy(x.T).pin_memory()
Out[21]:

 0  1
 2  3
 4  5
 6  7
 8  9
[torch.LongTensor of size 5x2]

In [22]: torch.from_numpy(np.ascontiguousarray(x.T)).pin_memory()
Out[22]:

 0  5
 1  6
 2  7
 3  8
 4  9
[torch.LongTensor of size 5x2]
```

**Comment by apaszke** [comment]
Pin memory is missing a `contiguous()` call inside. Should be an easy fix

**Comment by soumith** [comment]
fixed via #3273
